lib/cloudhsm/cloudhsm_key_generator.rb

```diff
@@ -0,0 +1,70 @@
+require 'greenletters'
+
+class CloudhsmKeyGenerator
+  KEY_MGMT_UTIL = '/opt/cloudhsm/bin/key_mgmt_util'.freeze
+  SSL_CONF = 'config/openssl.conf'.freeze
+
+  def initialize
+    @kmu = Greenletters::Process.new(KEY_MGMT_UTIL, transcript: $stdout)
+    @kmu.start!
+    @kmu.wait_for(:output, /Command:/)
+  end
+
+  def generate_saml_key
+    saml_label = create_key_and_crt_files
+    login_to_hsm
+    wrapping_key_handle = generate_symmetric_wrapping_key
+    import_private_key(saml_label, wrapping_key_handle)
+    exit_hsm
+    saml_label
+  end
+
+  private
+
+  def create_key_and_crt_files
+    timestamp = Time.zone.now.strftime('%Y%m%d%H%M%S')
+    saml_label = Shellwords.shellescape("saml_#{timestamp}")
+    system "openssl req -x509 -nodes -sha256 -days 365 -newkey rsa:2048 " \
+           "-keyout #{saml_label}.key " \
+           "-out #{saml_label}.crt -config #{SSL_CONF}"
```

Comment:
I would normally use `Subprocess.check_call` here instead of `system` so you get an error if the command fails and don't have to worry about escaping shell arguments.

Reply:
This process could use more attention to the private key's security. Are we intending to keep a paper backup of the SAML key? Then we should make that an explicit part of the process and delete the key when finished. It would also be better to store the key on a ramdisk or use encryption at rest so it never touches the instance's persistent disk in plaintext.

If we're not intending to keep a paper backup, then it would be better to generate the key inside the HSM rather than importing it.

Reply:
Are you sure? I'm pretty sure you can always export an asymmetric public key:

https://docs.aws.amazon.com/cloudhsm/latest/userguide/manage-keys.html#export-public-keys